pandas/core/indexes/base.py

```diff
@@ -251,6 +252,9 @@ def __new__(cls, data=None, dtype=None, copy=False, name=None,
         if name is None and hasattr(data, 'name'):
             name = data.name
 
+        if name is not None and not is_hashable(name):
+            raise TypeError('Index.name must be a hashable type')
+
```

Comment:
this very likely also needs checking for MultiIndex (as that's a different path in some cases).

Reply:
Do we allow non-hashable names for MultiIndex?

Reply:
no

Reply:
For a MultiIndex, it seems that `names` is converted into FrozenList after creation. I found [this answer](https://stackoverflow.com/questions/25676107/why-is-frozenlist-different-from-tuple) from you on StackOverflow about hashability of a FrozenList.

Right now, if `names` can’t be converted to a FrozenList, it throws an exception. For example:
```
In [1]: pd.MultiIndex(levels=[[1, 2], [u'one', u'two']],
   ...:                    labels=[[0, 0, 1, 1], [0, 1, 0, 1]],
   ...:                    names=(['foo'], ['bar']))
   ...:           

TypeError: unhashable type: 'list'
```
while this passes:
```
In [2]: pd.MultiIndex(levels=[[1, 2], [u'one', u'two']],
    ...:                    labels=[[0, 0, 1, 1], [0, 1, 0, 1]],
    ...:                    names=[('foo'), ('bar')])
```
Do we need to change anything here?

Reply:
no you just need to check that each name is hashable, not the frozen list itself. that's why .set_names is the best place for this